File: apps/web/lib/replay.ts
import { loadText } from 'shared/crdt';

/**
 * Op-log replay planning (Faz 6). Pure — no DB — so the reconstruction logic is
 * unit-testable in isolation. Given the persisted snapshots + surviving update
 * rows, compute the state as of a target update id (event-sourcing "state at K").
 *
 * INVARIANT #3 is why replay has a floor: the op log is pruned once a covering
 * snapshot is committed, so per-update history only survives *after* the latest
 * snapshot boundary. Targets whose granular updates were pruned are not exactly
 * replayable — planReplay returns null for those rather than lying.
 */

export interface SnapshotRow {
  upToUpdateId: bigint;
  state: Uint8Array;
}

export interface UpdateRow {
  id: bigint;
  update: Uint8Array;
}

export interface ReplayPlan {
  snapshot: Uint8Array | null;
  updates: Uint8Array[];
}

/**
 * Select the snapshot + ordered updates needed to reconstruct state as of just
 * after update row `target` (target === 0n → the base snapshot state, or empty).
 * Returns null when `target` falls inside the pruned window (granular history gone).
 */
export function planReplay(
  snapshots: SnapshotRow[],
  updates: UpdateRow[],
  target: bigint,
): ReplayPlan | null {
  // Newest snapshot that does not go past the target.
  const covering = snapshots
    .filter((s) => s.upToUpdateId <= target)
    .sort((a, b) => (a.upToUpdateId < b.upToUpdateId ? -1 : 1));
  const snapshot = covering.at(-1) ?? null;
  const base = snapshot ? snapshot.upToUpdateId : -1n;

  const selected = updates
    .filter((u) => u.id > base && u.id <= target)
    .sort((a, b) => (a.id < b.id ? -1 : 1));

  // Pruned floor: a snapshot exists beyond target but none covers up to it, and no
  // surviving update reaches it → its granular state is gone. Not replayable.
  if (!snapshot && selected.length === 0 && target > 0n) {
    const prunedPast = snapshots.some((s) => s.upToUpdateId > target);
    if (prunedPast) return null;
  }

  return { snapshot: snapshot?.state ?? null, updates: selected.map((u) => u.update) };
}

/** Reconstruct the plain text as of update `target`, or null if not replayable. */
export function replayText(
  snapshots: SnapshotRow[],
  updates: UpdateRow[],
  target: bigint,
): string | null {
  const plan = planReplay(snapshots, updates, target);
  if (!plan) return null;
  return loadText(plan.snapshot, plan.updates);
}

File: services/api-go/internal/rest/replay.go
package rest

import (
	"github.com/Contictus/collab-editor/services/api-go/internal/crdt"
)

// Replay planning (F6-10), mirroring replay.ts planReplay/replayText — pure,
// no DB. Given persisted snapshots + surviving update rows, compute the state
// as of just after update row target (0 → base snapshot state, or empty).
//
// Invariant #3 is why replay has a floor: pruned granular history is gone,
// and PlanReplay returns nil for it rather than lying.

// SnapshotRow mirrors SnapshotRow (ids as int64; JSON renders strings).
type SnapshotRow struct {
	UpToUpdateID int64
	State        []byte
}

// UpdateRow mirrors UpdateRow.
type UpdateRow struct {
	ID     int64
	Update []byte
}

// ReplayPlan is the snapshot + ordered updates to rebuild.
type ReplayPlan struct {
	Snapshot []byte
	Updates  [][]byte
}

// PlanReplay selects the newest snapshot not past target plus surviving
// updates in (base, target]. Nil when target falls in the pruned window.
func PlanReplay(snapshots []SnapshotRow, updates []UpdateRow, target int64) *ReplayPlan {
	var covering *SnapshotRow
	for i := range snapshots {
		if snapshots[i].UpToUpdateID <= target {
			if covering == nil || snapshots[i].UpToUpdateID > covering.UpToUpdateID {
				covering = &snapshots[i]
			}
		}
	}
	var snapshot []byte
	base := int64(-1)
	if covering != nil {
		snapshot = covering.State
		base = covering.UpToUpdateID
	}
	var selected [][]byte
	for _, u := range updates {
		if u.ID > base && u.ID <= target {
			selected = append(selected, u.Update)
		}
	}
	// Callers pass id-ordered rows (store order); selection preserves it.
	if snapshot == nil && len(selected) == 0 && target > 0 {
		for _, s := range snapshots {
			if s.UpToUpdateID > target {
				return nil
			}
		}
	}
	return &ReplayPlan{Snapshot: snapshot, Updates: selected}
}

// ReplayText reconstructs the plain text as of target, or false if not replayable.
func ReplayText(snapshots []SnapshotRow, updates []UpdateRow, target int64) (string, bool) {
	plan := PlanReplay(snapshots, updates, target)
	if plan == nil {
		return "", false
	}
	text, err := crdt.LoadText(plan.Snapshot, plan.Updates)
	if err != nil {
		return "", false
	}
	return text, true
}
